### PR TITLE
Transform dialect bringup

### PR DESCRIPTION
**Pull request by nicolasvasilache**
Bringup a generalized form of the transform dialect:

* Transform Dialect Bringup Branch
* Improve tests/transform_dialect/cuda/reduction.mlir by making input a… (#11208)
* [NFC] Refactor transform dialect matcher. (#11211)
* Add back mistakenly removed execution test in matmul CPU (#11209)
* [DONOTMERGETOMAIN] Drop a bunch of CI stuff we do not care about for … (#11215)
* Move transform dialect builder file to Common (#11216)
* Move the CPU to not use attributes. (#11219)
* NFC - Apply minor API cleanups (#11238)
* Add basic structured op matcher with fluent API (#11237)
* Add another cuda reduction test with other sizes and always tiling th… (#11243)
* Matcher API for inputs and inits/outputs (#11240)
* Support reduction+elementwise in transform dialect strategies (#11288)
* Refactor the matching code and move all the parameters into structs (#11222)
* [DONOTMERGETOMAIN] Add a flag to disable tensor fusion in flow (#11314)
* Add strategy+test for reduction(elementwise) (#11315)
* Add matcher callback operations to the transform dialect (#11389)
* Fixes related to IREEEraseHALDescriptorTypeFromMemRefOp
* Revert "[DONOTMERGETOMAIN] Add a flag to disable tensor fusion in flow (#11314)"
* Revert "[DONOTMERGETOMAIN] Drop a bunch of CI stuff we do not care about for … (#11215)"
* Revert "Transform Dialect Bringup Branch"
* Drop tests using --iree-disable-fusion-of-tensor-ops
* Use C++ callbacks to match reductions inside the transform dialect (#11390)

Co-authored-by: Oleksandr "Alex" Zinenko <zinenko@google.com>
Co-authored-by: Nicolas Vasilache <nicolas.vasilache@gmail.com>
Co-authored-by: Thomas Raoux <thomasraoux@google.com>
Co-authored-by: Matthias Springer <springerm@google.com>

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:161]: Took a while to decode `TFD` as `TileFuseAndDistribute`. Please add comment for that. Also is the templated form necessary.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:174]: Templatization and operator overloads... I am no C++ expert, so I am uncomfortable giving specific feedback, but this seems a bit hard to follow for vanilla C++ programmers like me.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:376]: Nit:  Please put TODOs before the statement...

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:1037]: I am assuming everything here is to be moved upstream? If so can we add these to `llvm-external-projects`?

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp:20]: This too. Lot of these seem like they are to be upstreamed.... If so can we land these in `llvm-external-projects`.

**Review by MaheshRavishankar — COMMENTED** [review]
Skimming through, dont find any major blockers here. Though this is very C++ heavy... I'd hope someone with more C++ knowledge can give better feedback here. Have a few questions about whether somethings are better suited to land in `llvm-external-projects`?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:174]: I think `Template Metaprogramming` is good for some cases (like allowing flexible and performant code for std::tuple, std::function, etc). But this is not the case in `buildTFDWithTileSizes`. I looked into the difference and found that one works on `ArrayRef<int64_t>` and the other work on `Value`. Can we avoid template metaprogramming for this case? Throwing out random ideas

1. Can we merge the two `buildTFDWithTileSizes` function? , can we make it `ArrayRef<OpFoldResult>` and force the callers to pass `ArrayRef<OpFoldResult>`?
2. Can we add an Option-like struct (like `SCFTilingOptions`, so we can set up the tile sizes through the Option, and pass it to `buildTFDWithTileSizes`?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:464]: style nit: naming with a leading "k" followed by mixed case.

```suggestion
static constexpr unsigned kCudaWarpSize = 32;
```

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:466]: This is a meaningless comment, please update it or remove it.

**Review by hanhanW — COMMENTED** [review]
I skimmed through the code, and just had few nits.

The major one is about template metaprogramming. Code that uses templates in complicated ways is often unreadable, and is hard to debug or maintain. It's fine to use it if they are used in a small number of low level components.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:80]: Why don't we just do `using mlir` and `using mlir::iree_compiler::IREE`?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:50]: nit: order?
   1st: #define
   2nd using
   3rd namespace mlir ...
?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:108]: Add doc? Missing context, but could we use an enum here instead?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:99]: and return?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:147]: nit: (void)create?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:161]: `buildTileFuseDistWithTileSizes`? :)

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:164]: `SmallVector<Value> *` -> `SmallVectorImpl<Value> *`? Same below

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:280]: can we use structured bindings here for std::tie?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:456]: ned?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:492]: Use a scalar?

**Review by dcaballe — COMMENTED** [review]
I reviewed up to `CommonExtensions.cpp`. I will follow up later!

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:280]: can we?
Not sure how to do this for variables that are already declared.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:99]: meh .. 2 more lines

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:80]: Dropped the ::mlir:: prefix.
I think we still want to avoid hiding the iree_compiler::IREE:: prefix and avoid conflation with the mlir:: prefix

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:1037]: Thomas proposed to move stuff as a followup to help distribute the load.
Leaving that to him, thanks Thomas!

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp:20]: Thomas proposed to move stuff as a followup to help distribute the load.
Leaving that to him, thanks Thomas!

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:297]: I think structured bindings can only be allowed with `auto`, i.e., `auto [a, b, c] = std::make_tuple(1, 2 3);`. In this case, `std::tie` should be used.

https://stackoverflow.com/questions/46764614/why-structured-bindings-only-work-with-auto

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:531]: This is inserting the transform dialect strategy within parent scope. So there is an implicit assumption that the pass itself is run on the parent scope. We can either pass the module in here explicitly to signify that the module is the translation unit, or we can insert it within the module, and discard it at the end of code-generation?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Not looking at the inner workings of the matchers/strategy builders etc. I am guessing these need to be upstreamed, etc. over time and people with more informed opinions will comment on these....  (Please do prioritize upstreaming it till we start making this load bearing in IREE).

+1 for this integration approach though... I have one feedback on where the generated transform dialect lives. Just marking request changes for that. I will remove my blocker before I go out on vacation, and let others chime in on the details.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp]: Changed the code to follow your second suggestion, we now insert within the module and discard it in the dropSchedule pass.

File: integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp
Comment (ThomasRaoux) [reply integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:255]: yes having a `transform.module` would be cleaner indeed. I'll let Nicolas comment on this.

**Comment by MaheshRavishankar** [comment]
I did say please wait for @dcaballe and @hanhanW to finish their reviews... its generally expected that everyone who has commented on the PR/ has opinions on the PR should have reviewed the changes.... especially for PRs like this

**Comment by benvanik** [comment]
Did the DONOTMERGETOMAIN commits get merged?

**Comment by MaheshRavishankar** [comment]
There were no changes to Flow AFAICS.

**Comment by benvanik** [comment]
+1 to waiting for reviewers for big branch merges like this - a warning that we're more likely to push back against this development style in the future (lightly reviewed or unreviewed branches that get merged to main without all reviewers fully getting a chance to participate).

**Comment by hanhanW** [comment]
I was planning to review it for a couple rounds because it is a very big PR. I can still leave some comments here and we can fix it afterwards...

**Comment by benvanik** [comment]
I'd prefer a revert - it's clean right now and will only get harder after integrates/etc.

**Comment by MaheshRavishankar** [comment]
If this wasnt a big PR, and just results in a lot of churn, I would have voted for revert too... but at the very least, please address any other comments by @dcaballe and @hanhanW before any further PRs on merging the transform dialect. I think in future, landing PRs without reviewers looking at it fully are grounds for revert.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:405]: The code below this line should be factored out to LLVMCPU/ and LLVMGPU/. They are not common utils, but they live in Coomon/. Some of the codes are CPU specific and some of them are CUDA specific.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp:20]: This is a cpp file, should we just using namespace `transform_dialect` or wrap them with the namespace, so we don't need `transform_dialect::` for every function implementation?

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.h:435]: ditto, it looks like GPU specific method, so it should not live in Common/.

**Review by hanhanW — COMMENTED** [review]
I took a deeper look, and think there are two things we can improve:

1. Wrap more things to the Option struct, so we don't need that much template.
2. Some CPU and GPU specific methods are living in Common/. We should leave the common utils, and move other parts to LLVMCPU/ and LLVMGPU/.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.h
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.h:435]: Trivial followup move that will happen independently as discussed above.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.cpp:20]: This is following MLIR best practices as most of the code will end up in a place closer to upstream when ready.
No idea where "wrapping in the namespace" comes from but it does not follow established practice.

**Comment by nicolasvasilache** [comment]
> ~Did the DONOTMERGETOMAIN commits get merged?~ nevermind looks like they were reverted?

Yes of course

**Comment by nicolasvasilache** [comment]
> I did say please wait for @dcaballe and @hanhanW to finish their reviews... its generally expected that everyone who has commented on the PR/ has opinions on the PR should have reviewed the changes.... especially for PRs like this

Then I'd suggest to either not click accept or disable the auto-merge button: github is oblivious to words.

File: integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp
Comment (nicolasvasilache) [reply integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:255]: What is `transform.module`?
Something IREE needs to know about or just a pretty printed name? 

If the former, I'll remind you that putting IR in a place that isn't just dropped by IREE is not trivial and requires deeper surgery that the IREE team need to decide on how to integrate in the custom nested pass pipeline.

If you're referring to something else, please elaborate into something actionnable.

**Comment by MaheshRavishankar** [comment]
> > I did say please wait for @dcaballe and @hanhanW to finish their reviews... its generally expected that everyone who has commented on the PR/ has opinions on the PR should have reviewed the changes.... especially for PRs like this
> 
> Then I'd suggest to either not click accept or disable the auto-merge button: github is oblivious to words.

Hmm, I didnt see the auto merge enabled. Something to keep in mind for future. Down side of auto merge feature. Maybe not good idea to have it on large PRs

File: integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp
Comment (MaheshRavishankar) [reply integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:255]: Just a way to disambiguate what is a module generated by transform dialect as opposed to something else. Works here cause it is removing an empty module (I think). Don't understand rest of the comment w.r.t surgery/ iree-team etc.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:188]: There are 4 parameters controlling the transformation, the suggestion is to splice one of them out into a struct for the purpose of removing one function call.

This is a bad tradeoff because distributing the control logic between 2 mechanisms (parameters and an options struct), loses uniformity and increases API confusion: we should use the same mechanism for everything.

Putting everything into a single options struct would be an anti-pattern for an MLIR build-like API and we should not be inventive here.

In fact, thinking about this deeper, I should have pushed back against the options struct to control fusion in the first place, the traditional MLIR solution here is more builders and more parameters.

> (It's always good to reduce the number of function arguments if there are many function arguments and the order matters, see https://abseil.io/tips/173)

It's really not when it goes against established practices and uniformity with the code around you. 
The MLIR/LLVM coding standard is what we should follow.
There may be a case to be made here to the MLIR community as to how to improve these things globally and uniformly.
This will involve deeper Tablegen surgery so I am not going to hold my breath on that :)

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:214]: discussed above

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:405]: This falls into the category of trivial things that take time and can be separated in a followup commit as was proposed by @ThomasRaoux .

Is this suggestion intended to be a blocker for landing or is a followup move-only PR the way forward?

File: integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp
Comment (nicolasvasilache) [reply integrations/tensorflow/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreter.cpp:255]: > Maybe an indication that transform dialect schedule needs to live in its own module transform.module, or something.

I do not understand what the suggestion is, so I can only speculate based on our prior discussions.

To make things concrete: are you suggesting something a first class module that lives in the IR somewhere e.g. within  the `hal.variant` / `hal.executable` op ? E.g.
```
hal.executable public @foo {
  hal.executable.variant @xxxx ... {
    hal.executable.export public @bar {
    ^bb0(%arg0: !hal.device, %arg1: index, %arg2 : index):
      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2
      hal.return %x, %y, %z : index, index, index
    }

    builtin.module { ... }

    builtin.module  {
      func.func @matmul_256x1024x128_div_sub() {
        ...
      }
    }
````

Or are you suggesting the transform dialect should have its own module op ?

Or something else ?

**Comment by nicolasvasilache** [comment]
> I took a deeper look, and think there are two things we can improve:

I believe I addressed that API-related comments in #11458.
Moving code around and into llvm-external-project can be done as a separate followup.

**Comment by hanhanW** [comment]
> > I took a deeper look, and think there are two things we can improve:
> 
> I believe I addressed that API-related comments in #11458. Moving code around and into llvm-external-project can be done as a separate followup.

Thanks for the detailed and explanation, and addressing the comments! The patch looks good to me!

Yeah, I agree that moving code around can be done as a separate follow-up, since it's a big PR (which already contains a lot of commits). Do we plan to do it after merging more things to main branch?

**Pull request merged**
